multi/stm32l4-multi/tty.c

```diff
@@ -99,60 +145,209 @@ static int tty_irqHandler(unsigned int n, void *arg)
 {
 	tty_ctx_t *ctx = (tty_ctx_t *)arg;
 
-	if (*(ctx->base + isr) & ((1 << 5) | (1 << 3))) {
+	if ((*(ctx->base + isr) & ((1 << 5) | (1 << 3))) != 0) {
 		/* Clear overrun error bit */
 		*(ctx->base + icr) |= (1 << 3);
 
-		lf_fifo_push(&ctx->rx_fifo, *(ctx->base + rdr));
-		ctx->rxready = 1;
+		lf_fifo_push(&ctx->data.irq.rx_fifo, *(ctx->base + rdr));
+		ctx->data.irq.rxready = 1;
 	}
 
-	if (tty_txready(ctx))
+	if (tty_txready(ctx) != 0) {
 		*(ctx->base + cr1) &= ~(1 << 7);
+	}
+
+	return 1;
+}
+
+
+/* Nonnegative type must be one of dma_ht or dm_tc,
+   negative type means idle line detected in which case it's negated number of bytes read. */
+static void tty_dmaCallback(void *arg, int type)
+{
+	tty_ctx_t *ctx = (tty_ctx_t *)arg;
+	int rxready = TTY_RX_READY_FLAG;
+
+	if (type == dma_tc) {
+		rxready |= TTY_RX_FILLED_FLAG;
+		atomic_fetch_add_explicit(&ctx->data.dma.debug.fill_cnt, 1, memory_order_acq_rel);
+	}
+
+	atomic_fetch_or_explicit(&ctx->data.dma.rxready, rxready, memory_order_acq_rel);
+}
+
+
+static int tty_irqHandlerDMA(unsigned int n, void *arg)
+{
+	tty_ctx_t *ctx = (tty_ctx_t *)arg;
+	int read;
+
+	/* Check for the idle line. */
+	if ((*(ctx->base + isr) & ((1 << 4))) == 0) {
+		return -1;
+	}
+	/* Clear idle line bit */
+	*(ctx->base + icr) |= (1 << 4);
+
+	read = libdma_leftToRx(ctx->data.dma.per);
+
+	if (read != 0) {
+		tty_dmaCallback(ctx, -read);
+	}
 
 	return 1;
 }
 
 
+static void tty_dmaHandleRx(tty_ctx_t *ctx)
+{
+	unsigned int write_pos, read_pos;
+	int wake = 0, wake_helper;
+	int rxready = atomic_exchange_explicit(&ctx->data.dma.rxready, TTY_RX_NOT_READY, memory_order_acq_rel);
+	unsigned int fill_cnt;
+	unsigned int rx_bytes;
+
+	if (rxready == TTY_RX_NOT_READY) {
+		return;
+	}
+
+	if ((rxready & TTY_RX_FILLED_FLAG) != 0) {
+		fill_cnt = atomic_exchange_explicit(&ctx->data.dma.debug.fill_cnt, 0, memory_order_acq_rel);
+		if (fill_cnt > 1) {
+			ctx->data.dma.debug.dropped_bytes += (fill_cnt - 1) * ctx->data.dma.rxbufsz;
+		}
+	}
+
+	libtty_putchar_lock(&ctx->tty_common);
+
+	write_pos = (ctx->data.dma.rxbufsz - libdma_leftToRx(ctx->data.dma.per)) & ctx->data.dma.rxbufsz_mask; /* Read once to prevent starvation of tx. */
```

Comment:
`rxready` and `leftToRx`  are read separately. This means leftToRx can return value after counter is reloaded on TC without (rxready & TTY_RX_FILLED_FLAG) being true.

Reply:
Move read of liftToTx to isrHandler